src/session_pool/session.js

```diff
@@ -0,0 +1,171 @@
+import EventEmitter from 'events';
+
+import { cryptoRandomObjectId } from 'apify-shared/utilities';
+import moment from 'moment';
+
+import { checkParamOrThrow } from 'apify-client/build/utils';
+import EVENTS from './events';
+
+
+/**
+ *  Class aggregating data for `Session`.
+ *  Session internal state can be enriched with custom user data for example some authorization tokens.
+ */
+export default class Session {
+    constructor(options = {}) {
+        const {
+            id = cryptoRandomObjectId(),
+            cookies = [],
+            fingerprintSeed = cryptoRandomObjectId(),
+            maxAgeSecs = 3000,
+            userData = {},
+            maxErrorScore = 3,
+            errorScoreDecrement = 0.5,
+            createdAt = moment().toISOString(),
+            usageCount = 0,
+            errorScore = 0,
+            maxSessionUsageCount = 50,
+            sessionPool,
+        } = options;
+
+        let { expiresAt } = options;
+
+        // Validation
+        checkParamOrThrow(id, 'options.id', 'Maybe String');
+        checkParamOrThrow(cookies, 'options.cookies', 'Maybe Array');
+        checkParamOrThrow(fingerprintSeed, 'options.fingerprintSeed', 'Maybe String');
+        checkParamOrThrow(maxAgeSecs, 'options.maxAgeSecs', 'Maybe Number');
+        checkParamOrThrow(userData, 'options.userData', 'Maybe Object');
+        checkParamOrThrow(maxErrorScore, 'options.maxErrorScore', 'Maybe Number');
+        checkParamOrThrow(expiresAt, 'options.expiresAt', 'Maybe String');
+        checkParamOrThrow(createdAt, 'options.createdAt', 'Maybe String');
+        checkParamOrThrow(usageCount, 'options.usageCount', 'Maybe Number');
+        checkParamOrThrow(errorScore, 'options.errorScore', 'Maybe Number');
+        checkParamOrThrow(maxSessionUsageCount, 'options.maxSessionUsageCount', 'Maybe Number');
+        checkParamOrThrow(sessionPool, 'options.sessionPool', 'Object');
+
+        // sessionPool must be at least instance of EvenEmitter.
+        // That way we can allow custom implementation of SessionPool in the future (It should not be needed).
+        if (!(sessionPool instanceof EventEmitter)) {
+            throw new Error('Session: sessionPool must be instance of SessionPool');
+        }
+
+        if (!expiresAt) {
+            expiresAt = moment().add(maxAgeSecs, 'seconds').toISOString();
```

Comment:
Why iso string and not Date object? You could use 

`this.expiresAt >= now` instead of

`moment(this.expiresAt).isSameOrBefore(now);`